model/strand.rb

```diff
@@ -40,14 +40,20 @@ def take_lease_and_reload
           fail "BUG: strand with @deleted set still exists in the database"
         end
       else
-        num_updated = DB[<<SQL, id, lease_time].update
+        DB.transaction do
+          lease_clear_debug_snapshot = this.for_update.all
```

Comment:
Does `for_update` not lock the row until the transaction is committed?

We're attempting to detect lease violations and generate more detailed logs. However, if it locks the strand, the intruder will simply wait for the lock, thus avoiding any lease violation.

This might not be the case, but I'm seeking clarification to ensure my understanding is correct.

Reply:
> Does `for_update` not lock the row until the transaction is committed?

It locks the row until the transaction is committed, which means it doesn't do anything if you are in autocommit (no bracketed BEGIN/COMMIT). That's why I shove the DB.transaction block around it. 

> We're attempting to detect lease violations and generate more detailed logs. However, if it locks the strand, the intruder will simply wait for the lock, thus avoiding any lease violation.

Not sure what you mean. I want to dump the state of the record matching the strand ID alone, before going on to run the normal lease-freeing code. I do not want a record mutation in between getting the copy of the record and running the lease-release code.

The crux of this patch is that the lease-release code is an UPDATE statement _with a filter_. It's ambiguous whether a strand has been deleted entirely (i.e. a `id`-only match would also yield no records) or if the `lease` attribute has an unexpected value. That's why I can't add a few more fields to `RETURNING` and get the information I want.